MonstieWash/Assets/Scripts/UI/ProgressBarUI.cs

```diff
@@ -0,0 +1,56 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+using System.Linq;
+
+public class ProgressBarUI : MonoBehaviour
+{
+    private GameSceneManager m_gameSceneManager;
+    private RectMask2D m_progressMask;
+    private RectTransform m_maskTransform;
+    private List<TaskData> m_taskData;
+    private float m_sceneProgress = 0;
+    private int m_taskCount = 0;
+
+    private void Awake()
+    {
+        m_gameSceneManager = FindFirstObjectByType<GameSceneManager>();
+        m_progressMask = GetComponent<RectMask2D>();
+        m_maskTransform = GetComponent<RectTransform>();
+        m_taskData = FindFirstObjectByType<TaskTracker>().TaskData;
+    }
+
+    private void OnEnable()
+    {
+        m_gameSceneManager.OnSceneChanged += CheckCompletion;
+    }
+
+    private void OnDisable()
+    {
+        m_gameSceneManager.OnSceneChanged -= CheckCompletion;
+    }
+
+/// <summary>
+/// Recalculates tasks completed and total tasks for the current active scene
+/// </summary>
+    public void CheckCompletion()
+    {
+        m_sceneProgress = 0f;
+        m_taskCount = 0;
+        foreach (var task in m_taskData.Where(t => t.gameObject.scene == m_gameSceneManager.CurrentScene))
```

Comment:
literally the only thing is to use the `String.Equals()` method when working with strings since `==` can be unreliable.

```suggestion
        foreach (var task in m_taskData.Where(t => t.gameObject.scene.Equals(m_gameSceneManager.CurrentScene)))
```

Reply:
With this, because im comparing Scene objects and not strings, is it still preferable to use String.Equals(). And also is there a benefit to comparing scenes using names/strings vs scene objects? Just trying to pick your brain for some knowledge.

Reply:
Oh lol I'm just blind then whoopsies. Dw about that then.

I had a look into the unity code for scenes, and turns out each scene has a unique identifier for each instance that it uses to do equality checks, and the identifier is just an int. This means that I think comparing scenes the way you did above is actually faster, so disregard my suggestion. I'll go back and look at changing the other scene comparisons in the codebase to do it this way at some point as well.